transport/cluster.go

```diff
@@ -3,42 +3,113 @@ package transport
 import (
 	"fmt"
 	"log"
+	"strconv"
 	"time"
 
 	"scylla-go-driver/frame"
 	. "scylla-go-driver/frame/response"
 
+	"github.com/google/btree"
 	"go.uber.org/atomic"
 )
 
 type (
-	PeerMap = map[string]*Node
+	PeerMap         = map[string]*Node
+	DCRacksMap      = map[string]int
+	DCRepFactorsMap = map[string]repFactor
+	KSMap           = map[string]keyspace
 
 	requestChan chan struct{}
+
+	repFactor uint32
 )
 
+const BTreeDegree = 2
+
 type Cluster struct {
-	peers             atomic.Value // PeerMap
+	topology          atomic.Value // *topology
 	control           *Conn
 	cfg               ConnConfig
 	handledEvents     []frame.EventType // This will probably be moved to config.
 	knownHosts        []string
+	keyspace          string
 	refreshChan       requestChan
 	reopenControlChan requestChan
 	closeChan         requestChan
+}
+
+type topology struct {
+	peers     PeerMap
+	dcRacks   DCRacksMap
+	allNodes  []*Node
+	ring      btree.BTree
+	keyspaces KSMap
+}
 
-	// Yet to be used.
-	// ring        map[Token]*Node
-	// dataCenters map[string]*DataCenter
+type keyspace struct {
+	stg strategy
+	// TODO: Add and use attributes below
+	// tables		      map[string]table
+	// user_defined_types map[string](string, cqltype)
 }
 
-// type dataCenter struct {
-// 	nodes   []Node
-// 	rackCnt uint // Type?
-// }
+type stgClass int
+
+const (
+	networkTopology stgClass = iota
+	simple
+	local
+	other
+)
+
+type strategy struct {
+	class stgClass
+	// used in simple strategy
+	rf repFactor
+	// used in networkTopology strategy
+	dcRF DCRepFactorsMap
+	// used in other strategy
+	name string
+	data map[string]string
+}
+
+type QueryInfo struct {
+	token    *Token
+	topology *topology
+	stg      strategy
+}
+
+func (c *Cluster) newQueryInfo(t *Token) QueryInfo {
+	top := c.topology.Load().(*topology)
+	return QueryInfo{
+		token:    t,
+		topology: top,
+		stg:      top.keyspaces[c.keyspace].stg, // TODO: do we assume, that keyspace is always specified?
+	}
+}
+
+// ringRange return slice of nodes (desirably of length cnt) holding data described by token.
+// wanted function allows applying additional requirements for nodes to be taken.
+func (t *topology) ringRange(token Token, cnt int, wanted func(*Node, []*Node) bool) []*Node {
+	replicas := make([]*Node, 0, cnt)
+	it := func(i btree.Item) bool {
+		n := i.(RingEntry).node
+		if wanted(n, replicas) {
```

Comment:
why would we pass replicas here?

Reply:
Token aware strategies make decisions about taking specific node based on already taken nodes.
Since Simple strategy only cares about choosing unique nodes (which perhaps is not even necessary, because we believe in Scylla and the chances of getting incomplete information about topology is rather slim), maybe we should give up passing **wanted** function and just hard code it inside **ringRange** (name will be changed)?